Add painless method getByPath, get value from nested collections with dotted path

stu-elastic opened the pull request:
Given a nested structure composed of Lists and Maps, getByPath will return the value
keyed by path.  getByPath is a method on Lists and Maps.

The path is string Map keys and integer List indices separated by dot. An optional third
argument returns a default value if the path lookup fails due to a missing value.

Eg.
['key0': ['a', 'b'], 'key1': ['c', 'd']].getByPath('key1') = ['c', 'd']
['key0': ['a', 'b'], 'key1': ['c', 'd']].getByPath('key1.0') = 'c'
['key0': ['a', 'b'], 'key1': ['c', 'd']].getByPath('key2', 'x') = 'x'
[['key0': 'value0'], ['key1': 'value1']].getByPath('1.key1') = 'value1'

Throws IllegalArgumentException if an item cannot be found and a default is not given.
Throws NumberFormatException if a path element operating on a List is not an integer.

Fixes #42769

stu-elastic committed:
Add painless method getByPath, get value from nested collections with dotted path

Given a nested structure composed of Lists and Maps, getByPath will return the value
keyed by path.  getByPath is a method on Lists and Maps.

The path is string Map keys and integer List indices separated by dot. An optional third
argument returns a default value if the path lookup fails due to a missing value.

Eg.
['key0': ['a', 'b'], 'key1': ['c', 'd']].getByPath('key1') = ['c', 'd']
['key0': ['a', 'b'], 'key1': ['c', 'd']].getByPath('key1.0') = 'c'
['key0': ['a', 'b'], 'key1': ['c', 'd']].getByPath('key2', 'x') = 'x'
[['key0': 'value0'], ['key1': 'value1']].getByPath('1.key1') = 'value1'

Throws IllegalArgumentException if an item cannot be found and a default is not given.
Throws NumberFormatException if a path element operating on a List is not an integer.

Fixes #42769

elasticmachine commented:
Pinging @elastic/es-core-infra

stu-elastic committed:
Merge branch 'master' of github.com:elastic/elasticsearch into painless/42769-getByPath

rjernst commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/api/Augmentation.java:
```diff
@@ -552,4 +553,78 @@ public static String decodeBase64(String receiver) {
+     */
+    private static Object getByPath(Object receiver, String path, Object defaultValue) {
+        Object next = receiver;
+        String[] elements = path.split("\\.");
```
While an iterative solution does work, I think we can eliminate the double double casting necessary in the first call (needing to check instanceof on the receiver even though we just knew the type in the outer call).

Instead of splitting the entire path, I think there should be 2 internal methods, which take Map or List as the receiver, and have an additional argument, the current index in the path. You can then use `indexOf` along with that current index to find the location of the next dot, and extract the current path element.

So there would be a few methods:
* one taking Map, this would have the default calling logic based on whether the map contains the next element, or there is no next path element
* one taking List, this would have the default calling logic based on length of the list, or there is no next path element
* an internal dispatch method that would take Object, which the other two methods would call when they successfully extract the next element

stu-elastic replied:
Changed to recursion as requested.

rjernst commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/api/Augmentation.java:
```diff
@@ -552,4 +553,78 @@ public static String decodeBase64(String receiver) {
+     * not exist. Throws 'NumberFormatException' if an path element for a List is not an integer.
+     */
+    private static Object getByPath(Object receiver, String path) {
+        Object value = getByPath(receiver, path, new missingValue());
```
I think we can avoid having a special missingValue by passing in a Supplier as the default value to the primary methods, where here would would give an instance that throws an exception.

stu-elastic replied:
Now using a Supplier.

rjernst reviewed (commented):
I left a few comments. Overall I think we can make this cleaner by using recursion in this case, since we need to switch back and forth between types. You could still do this iteratively, but then I would have the first part of the loop do the extraction of the next element, and the second part do the dispatch of the next receiver. The downside is we would still have an extra instanceof check when we already know the receiver type from the whitelisted methods.

I also am wondering if we should be more strict in the default value behavior, but still allow for missing behavior. My suggestion would be to make this work strictly using the current dot syntax, and in a followup add support in the path syntax for the null-safe operator. This way a user can be very explicit if they want lenient behavior at any element within the path.

stu-elastic commented on modules/lang-painless/src/test/java/org/elasticsearch/painless/AugmentationTests.java:
```diff
@@ -232,11 +235,134 @@ public void testString_SplitOnToken() {
             new SplitCase("1\n1.1.\r\n1\r\n111", "\r\n"),
         };
         for (SplitCase split : cases) {
-            //System.out.println(String.format("Splitting '%s' by '%s' %d times", split.input, split.token, split.count));
```
Good idea.

rjernst replied:
Oops, nevermind, I looked at this wrong. You are removing it as I suggested!

stu-elastic commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/api/Augmentation.java:
```diff
@@ -552,4 +553,78 @@ public static String decodeBase64(String receiver) {
+                    throw new NumberFormatException(String.format(Locale.ROOT, format, element, path, i));
+                }
+            } else {
+                return defaultValue;
```
A not a map or list while there are still path elements being processed.  What's the desired behavior?

rjernst replied:
I would throw an exception. This isn't necessarily parsed json, it could be called on any Object.

stu-elastic replied:
Throwing `IllegalArgumentException` now.

stu-elastic commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/api/Augmentation.java:
```diff
@@ -552,4 +553,78 @@ public static String decodeBase64(String receiver) {
+     * getByPath for Lists without default
+     */
+    public static <T> Object getByPath(List<T> receiver, String path) {
+        return getByPath((Object)receiver, path);
```
It's to avoid infinite recusion, without the cast it'd call itself rather than 
`private static Object getByPath(Object receiver, String path, Object defaultValue)`
on 573.

stu-elastic committed:
Implement getByPath using recursion.

Separate getByPath tests into different test methods rather than a single
  test driven by cases.

Raise IllegalArgumentException when object at non-terminal path element
  is not a container, even if defaultValue is given.

Check for double . in paths and paths ending in .

stu-elastic commented on modules/lang-painless/src/test/java/org/elasticsearch/painless/AugmentationTests.java:
```diff
@@ -232,11 +235,134 @@ public void testString_SplitOnToken() {
+
+        GetByPathCase[] cases = new GetByPathCase[] {
+          // basic
+          new GetByPathCase("['k0':'v0']", "k0", "v0"),
```
Moved into own class, used different methods for each group of cases and added the assert methods as suggested.

rjernst commented on modules/lang-painless/src/test/java/org/elasticsearch/painless/GetByPathAugmentationTests.java:
```diff
@@ -0,0 +1,183 @@
+        assertPathError(toScript(collection, key, defaultValue), error);
+    }
+
+    private void assertPathError(String script, IllegalArgumentException error) {
```
We don't need to reproduce the exception to check it. Just pass in the message here that we want to check for. You can even just check for a substring with the `containsString` matcher.

stu-elastic replied:
Just passing in the message now.

rjernst commented on modules/lang-painless/src/test/java/org/elasticsearch/painless/GetByPathAugmentationTests.java:
```diff
@@ -0,0 +1,183 @@
+        assertPathError("[['a','b'],['c','d']]", path, "'foo'", numberFormat("k0", path, 1));
+
+        path = "k0.k01.k012";
+        assertPathError(mapMapList, path, "'foo'", numberFormat("k012", path, 2));
```
Each one of these assertions can be in its own test method. This makes reproductions much smaller and easier to deal with when debugging. For example, then when adding breakpoints, this makes it simpler to not have to deal with skipping over other assertions that is not the one being debugged.

Naming-wise, I would try to have descriptive but concise names. So we don't need the "GetByValue" in every test name, since that is what this entire test class is about. As an example, for this line one I might call it `testNonIntListIndexWithDefaultFails()`. The test name is also a form of documentation for future readers here, with the name giving some indication of what is trying to be tested.

The test method names would be something like `test

stu-elastic replied:
That's much nicer, moved each to their own method.

rjernst commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/api/Augmentation.java:
```diff
@@ -552,4 +554,138 @@ public static String decodeBase64(String receiver) {
+     * Throws 'IllegalArgumentException' if elements[i] is an empty string, indicating double separators
+     * or path ends in a separator.
+     */
+    private static Object getByPath(Object obj, String[] elements, int i, Supplier<Object> defaultSupplier) {
```
I think it would be more clear with each of these names having one more word. Eg Dispatch here, Map/List below.

rjernst commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/api/Augmentation.java:
```diff
@@ -552,4 +554,138 @@ public static String decodeBase64(String receiver) {
+    /**
+     * List version of getByPath, calls dispatch function on next path element with object list[elements[i]] or
+     * return defaultSupplier if element is missing an i is the last index of elements.
+     * Throws 'IllegalArgumentException' is element is missing and i is not the last index of elements.
```
You can use `@throws` instead of prose here instead of text. Same for `@return` above

stu-elastic replied:
Updated to use javadoc `@throws` and `@return`.

rjernst commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/api/Augmentation.java:
```diff
@@ -552,4 +554,138 @@ public static String decodeBase64(String receiver) {
+    }
+
+    /**
+     * Same as 'getByPath(Map, String)' but returns defaultValue if path does not exist.
```
I think the main logic you describe needs to be moved up to these methods. There will be no javadocs generated for private methods.

Also, you can refer to other methods with `{@link #methodName(signature...)}`. So you can have most of the javadocs on one of the public methods, then link to that and describe the differences in the others.

stu-elastic replied:
Moved all docs to top level method and `{@link..}` to it from variants.

rjernst commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/api/Augmentation.java:
```diff
@@ -552,4 +554,138 @@ public static String decodeBase64(String receiver) {
+        }
+        String[] elements = path.split("\\.");
+        if (elements.length == 1 && elements[0].length() == 0) {
+            throw new IllegalArgumentException("Missing path");
```
We could check for this before actually splitting right? It would just be empty path?

stu-elastic replied:
Yup, moved that check up.

rjernst reviewed (commented):
This looks good, just a few more minor comments.

stu-elastic committed:
One test per method.

Rename helper functions.

Use javadoc for public functions.

Optimize empty path check.

stu-elastic committed:
Merge branch 'master' of github.com:elastic/elasticsearch into painless/42769-getByPath

stu-elastic committed:
Remove double newlines between methods

rjernst commented on modules/lang-painless/src/test/java/org/elasticsearch/painless/GetByPathAugmentationTests.java:
```diff
@@ -0,0 +1,259 @@
+        assertPathError(mapMap, "a..b", "Extra '.' in path [a..b] at index [1]");
+    }
+
+    public void testTrainingDot() {
```
typo? training -> trailing

stu-elastic replied:
Fixed.

rjernst reviewed (approved):
LGTM

stu-elastic committed:
type: testTrainingDot -> testTrailingDot

jdconrad reviewed (approved):
LGTM.  Thanks for adding this method Stuart!

stu-elastic committed:
Merge branch 'master' of github.com:elastic/elasticsearch into painless/42769-getByPath
